samples/nrf9160/nrf_cloud_rest_device_message/src/main.c

```diff
@@ -0,0 +1,379 @@
+		return err;
+	}
+
+	/* Send the message to nRF Cloud */
+	err = nrf_cloud_rest_send_device_message(&rest_ctx, device_id, msg, false, NULL);
+
+	if (err) {
+		LOG_ERR("Failed to send device message via REST: %d", err);
+		return err;
+	}
+
+	return 0;
+}
+
+int send_message_on_button(void)
+{
+	/* Wait for a buttonpress */
+	(void)k_sem_take(&button_press_sem, K_FOREVER);
+
+	/* Send off a JSON message about the buttonpress */
+	/* This matches the Button app schema. */
+	return send_message("{\"appId\":\"BUTTON\", \"messageType\":\"DATA\", \"data\":\"1\"}");
+}
+
+static int do_jitp(void)
+{
+	int ret = 0;
+
+	LOG_INF("Performing JITP...");
+	ret = nrf_cloud_rest_jitp(CONFIG_NRF_CLOUD_SEC_TAG);
+
+	if (ret == 0) {
+		LOG_INF("Waiting 30s for cloud provisioning to complete...");
+		k_sleep(K_SECONDS(30));
+		k_sem_reset(&button_press_sem);
+		LOG_INF("Associate device with nRF Cloud account and press button %d when complete",
+			BTN_NUM);
+		(void)k_sem_take(&button_press_sem, K_FOREVER);
+		return 0;
+	} else if (ret == 1) {
+		LOG_INF("Device already provisioned");
+		return 0;
+	}
+
+	LOG_ERR("JITP device provisioning failed: %d", ret);
+	return ret;
+}
+
+static int modem_time_wait(void)
+{
+	int err;
+	char time_buf[64];
+
+	LOG_INF("Waiting for modem to acquire network time...");
+
+	do {
+		k_sleep(K_SECONDS(3));
+		err = nrf_modem_at_cmd(time_buf, sizeof(time_buf), "AT%%CCLK?");
+	} while (err != 0);
+
+	if (err) {
+		LOG_ERR("AT Clock Command Error %d", err);
+		return err;
+	}
+
+	LOG_INF("Network time obtained");
+	return 0;
+}
+
+static void lte_handler(const struct lte_lc_evt *const evt)
+{
+	switch (evt->type) {
+	case LTE_LC_EVT_NW_REG_STATUS:
+		LOG_DBG("LTE_LC_EVT_NW_REG_STATUS: %d", evt->nw_reg_status);
+		if ((evt->nw_reg_status != LTE_LC_NW_REG_REGISTERED_HOME) &&
+		     (evt->nw_reg_status != LTE_LC_NW_REG_REGISTERED_ROAMING)) {
+			break;
+		}
+
+		LOG_DBG("Network registration status: %s",
+			evt->nw_reg_status == LTE_LC_NW_REG_REGISTERED_HOME ?
+			"Connected - home network" : "Connected - roaming");
+		k_sem_give(&lte_connected);
+		break;
+	case LTE_LC_EVT_CELL_UPDATE:
+		LOG_DBG("LTE cell changed: Cell ID: %d, Tracking area: %d",
+			evt->cell.id, evt->cell.tac);
+		break;
+	default:
+		break;
+	}
+}
+
+static int connect_to_lte(void)
+{
+	int err;
+
+	LOG_INF("Waiting for network...");
+
+	k_sem_reset(&lte_connected);
+
+	err = lte_lc_init_and_connect_async(lte_handler);
+	if (err) {
+		LOG_ERR("Failed to init modem, error: %d", err);
+		return err;
+	}
+
+	k_sem_take(&lte_connected, K_FOREVER);
+	(void)set_led(1);
+	LOG_INF("Connected to LTE");
+	return 0;
+}
+
+
+int setup_connection(void)
+{
+	/* Get the device ID */
+	int err = nrf_cloud_client_id_get(device_id, sizeof(device_id));
+
+	if (err) {
+		LOG_ERR("Failed to get device ID, error: %d", err);
+		return err;
+	}
+
+	LOG_INF("Device ID: %s", log_strdup(device_id));
+
+	/* Connect to LTE */
+	err = connect_to_lte();
+
```

Comment:
I think it helps readability to group related statements without a blank line between:
```
err = connect_to_lte();
if (err) {
...

Reply:
I agree, but I cannot do this on the first err = due to our coding style guidelines on variable declaration, and so I add the newline everywhere for consistency.

I could, alternatively, declare err at the beginning of the function, but that will lead to the potentially equally irksome situation of a definition right after a declaration in a few cases.

What do you think?

Reply:
I'd just declare err at the beginning in this case.

Reply:
I agree generally with declaring right before first use, but only in large functions.  Here it's not that far from the points of use.

I also agree to an extent with the objection to a newline after a variable declaration, but that's not a battle worth fighting (or likely to be won).